storm-webapp/src/main/java/org/apache/storm/daemon/logviewer/webapp/LogviewerResource.java

```diff
@@ -0,0 +1,268 @@
+        meterDownloadLogFileHttpRequests.mark();
+
+        String user = httpCredsHandler.getUserName(request);
+        String file = request.getParameter("file");
+        String decodedFileName = URLDecoder.decode(file);
+        return logDownloadHandler.downloadLogFile(decodedFileName, user);
+    }
+
+    /**
+     * Handles '/daemondownload' (downloading specific daemon log file) request.
+     */
+    @GET
+    @Path("/daemondownload")
+    public Response downloadDaemonLogFile(@Context HttpServletRequest request) throws IOException {
+        meterDownloadLogDaemonFileHttpRequests.mark();
+
+        String user = httpCredsHandler.getUserName(request);
+        String file = request.getParameter("file");
+        String decodedFileName = URLDecoder.decode(file);
+        return logDownloadHandler.downloadDaemonLogFile(decodedFileName, user);
+    }
+
+    /**
+     * Handles '/search' (searching from specific worker or daemon log file) request.
+     */
+    @GET
+    @Path("/search/{file}")
+    public Response search(@PathParam("file") String file, @Context HttpServletRequest request) throws IOException {
+        String user = httpCredsHandler.getUserName(request);
+        boolean isDaemon = StringUtils.equals(request.getParameter("is-daemon"), "yes");
+        String decodedFileName = URLDecoder.decode(file);
+        String searchString = request.getParameter("search-string");
+        String numMatchesStr = request.getParameter("num-matches");
+        String startByteOffset = request.getParameter("start-byte-offset");
+        String callback = request.getParameter("callback");
+        String origin = request.getHeader("Origin");
+
+        try {
+            return logSearchHandler.searchLogFile(decodedFileName, user, isDaemon, searchString, numMatchesStr,
+                    startByteOffset, callback, origin);
+        } catch (InvalidRequestException e) {
+            LOG.error(e.getMessage(), e);
+            return new JsonResponseBuilder().setData(UIHelpers.exceptionToJson(e)).setCallback(callback)
+                    .setStatus(400).build();
+        }
+    }
+
+    /**
+     * Handles '/deepSearch' request.
+     */
+    @GET
+    @Path("/deepSearch/{topoId}")
+    public Response deepSearch(@PathParam("topoId") String topologyId,
+                               @Context HttpServletRequest request) throws IOException {
+        String user = httpCredsHandler.getUserName(request);
+        String searchString = request.getParameter("search-string");
+        String numMatchesStr = request.getParameter("num-matches");
+        String portStr = request.getParameter("port");
+        String startFileOffset = request.getParameter("start-file-offset");
+        String startByteOffset = request.getParameter("start-byte-offset");
+        String searchArchived = request.getParameter("search-archived");
+        String callback = request.getParameter("callback");
+        String origin = request.getHeader("Origin");
```

Comment:
Just an FYI Headers and parameters can be marked with annotations that are pulled out as parameters.  I am not sure if this is cleaner or if that would be.  Bother are rather long.